pkg/ruler/rule_concurrency.go

```diff
@@ -187,6 +178,78 @@ func (c *TenantConcurrencyController) Allow(_ context.Context, group *rules.Grou
 	return false
 }
 
+func (c *TenantConcurrencyController) SplitGroupIntoBatches(_ context.Context, g *rules.Group) []rules.ConcurrentRules {
+	if !c.isGroupAtRisk(g) {
+		// If the group is not at risk, we can evaluate the rules sequentially.
+		return nil
+	}
+
+	logger := log.With(c.logger, "group", g.Name())
+
+	type ruleInfo struct {
+		ruleIdx                 int
+		unevaluatedDependencies map[rules.Rule]struct{}
+	}
+	remainingRules := make(map[rules.Rule]ruleInfo)
+	firstBatch := rules.ConcurrentRules{}
+	for i, r := range g.Rules() {
+		if r.NoDependencyRules() {
+			firstBatch = append(firstBatch, i)
+			continue
+		}
+		// Initialize the rule info with the rule's dependencies.
+		// Use a copy of the dependencies to avoid mutating the rule.
+		info := ruleInfo{ruleIdx: i, unevaluatedDependencies: map[rules.Rule]struct{}{}}
+		for _, dep := range r.DependencyRules() {
+			info.unevaluatedDependencies[dep] = struct{}{}
+		}
+		remainingRules[r] = info
+	}
+	if len(firstBatch) == 0 {
+		// There are no rules without dependencies.
+		// Fall back to sequential evaluation.
+		level.Info(logger).Log("msg", "No rules without dependencies found, falling back to sequential rule evaluation. This may be due to indeterminate rule dependencies.")
+		return nil
+	}
+	order := []rules.ConcurrentRules{firstBatch}
+
+	// Build the order of rules to evaluate based on dependencies.
+	for len(remainingRules) > 0 {
+		previousBatch := order[len(order)-1]
+		// Remove the batch's rules from the dependencies of its dependents.
+		for _, idx := range previousBatch {
+			rule := g.Rules()[idx]
+			for _, dependent := range rule.DependentRules() {
+				dependentInfo := remainingRules[dependent]
+				delete(dependentInfo.unevaluatedDependencies, rule)
+			}
+		}
+
+		var batch rules.ConcurrentRules
+		// Find rules that have no remaining dependencies.
+		for name, info := range remainingRules {
+			if len(info.unevaluatedDependencies) == 0 {
+				batch = append(batch, info.ruleIdx)
+				delete(remainingRules, name)
+			}
+		}
+
+		if len(batch) == 0 {
+			// There is a cycle in the rules' dependencies.
+			// We can't evaluate them concurrently.
+			// Fall back to sequential evaluation.
+			level.Warn(logger).Log("msg", "Cyclic rule dependencies detected, falling back to sequential rule evaluation")
+			return nil
+		}
+
+		order = append(order, batch)
+	}
+
+	level.Info(logger).Log("msg", "Batched rules into concurrent blocks", "rules", len(g.Rules()), "batches", len(order))
```

Comment:
Can we add a debug line that tells us exactly how the batches look like?

Reply:
Done! Looks like this:
```
{"batches":[[0,4,8,12,16,20,24,28,32,36,37,38,58],[39,42,18,30,56,33,41,45,6,9,10,21,26,46,13,17,29,1,51,55,14,52,34,40,2,5,22,25],[23,15,19,31,11,27,7,35,3],[43,44,54,49,50,57,47,48,53]],"component":"concurrency-controller","group":"test-group","level":"debug","msg":"Batched rules into concurrent blocks","tenant":"user1"}
```